tools/autograd/derivatives.yaml

```diff
@@ -305,15 +305,28 @@
 
 # For clamp, gradient is not defined at the boundaries. But empirically it's helpful
 # to be able to get gradient on min and max, so we return the subgradient 1 for these cases.
+- name: clamp.Tensor(Tensor self, Tensor? min=None, Tensor? max=None) -> Tensor
+  self: clamp_backward(grad, self, min, max)
+  min: "grad.defined() && min.defined() ? grad * (self < min).to(grad.dtype()) : Tensor{}"
```

Comment:
checking if grad is defined should not be needed here right? the codegen should handle that already.

Reply:
I was getting autograd test failures without it. Looking at the generated `ClampBackward0::apply` it looks like the complex grad expression doesn't automatically get the grad not defined check:
```cpp
  if (should_compute_output({ max_ix })) {
    auto grad_result = grad.defined() && max.defined() ? grad * (self > max).to(grad.dtype()) : Tensor{};
    copy_range(grad_inputs, max_ix, grad_result);
  }
  if (should_compute_output({ min_ix })) {
    auto grad_result = grad.defined() && min.defined() ? grad * (self < min).to(grad.dtype()) : Tensor{};
    copy_range(grad_inputs, min_ix, grad_result);
  }
  if (should_compute_output({ self_ix })) {
    auto grad_result = any_grad_defined ? (clamp_backward(grad, self, min, max)) : Tensor();
    copy_range(grad_inputs, self_ix, grad_result);
  }
```

Reply:
Based on the logic https://github.com/pytorch/pytorch/blob/3489b4a7b8629ce53b0bc6a3cf6fdc6107edd070/tools/autograd/gen_autograd_functions.py#L219-L228 it should be no? There is a single output and you define the formulas for each input one by one.
Am I missing something?

Reply:
`arg.argument.type` would be `'Tensor?'`, right?

Reply:
@albanD I could move these checks into codegen if you like. Something like
```python
if str(arg.argument.type) == 'Tensor?':
    formula = 'any_grad_defined && {arg.name}.defined() ? ({formula}) : Tensor{}'
```

Reply:
Ho yes that would be great! Thanks!